### ISSUE TITLE
Try to turn off / avoid using `eval()` as much as possible

### ISSUE THREAD

**Issue by swankjesse** [cashapp/zipline#646]
It’s a potential sandbox-escape if we end up running unexpected code.

**Comment by swankjesse** [comment]
Two different projects.

1. Avoid eval() in our libraries
2. Disable eval() in our runtime

**Comment by swankjesse** [comment]
For 1, I can’t find any calls to eval() in our generated JS. Maybe it was eliminated when we upgraded to Kotlin 1.7?

**Comment by JakeWharton** [comment]
It used to be used in development webpack but I believe we turned it off. Should never be used in production Kotlin unless you opt-in somehow or call it explicitly.

**Comment by swankjesse** [comment]
Yep, there’s zero use of eval in our existing Kotlin/JS compiler output.

**Comment by swankjesse** [comment]
I tried disabling eval. Disabling it is relatively straightforward – use `JS_NewContextRaw` and introduce only the intrinsics we want:

```
/**
 * Like JS_NewContext, but without an intrinsic for eval.
 */
JSContext* jsNewContextNoEval(JSRuntime* rt) {
  JSContext *ctx;

  ctx = JS_NewContextRaw(rt);
  if (!ctx) return NULL;

  JS_AddIntrinsicBaseObjects(ctx);
  JS_AddIntrinsicDate(ctx);
  // JS_AddIntrinsicEval(ctx); // TODO(jwilson): remove
  JS_AddIntrinsicStringNormalize(ctx);
  JS_AddIntrinsicRegExp(ctx);
  JS_AddIntrinsicJSON(ctx);
  JS_AddIntrinsicProxy(ctx);
  JS_AddIntrinsicMapSet(ctx);
  JS_AddIntrinsicTypedArrays(ctx);
  JS_AddIntrinsicPromise(ctx);
  // JS_AddIntrinsicBigInt(ctx);
  return ctx;
}
```

Unfortunately actually doing this is currently blocked by a few things:
 * `installFinalizationRegistry` uses `JS_Eval`. We could work around this by compiling first on a separate `JSContext`.
 * QuickJs.compile() uses `JS_Eval`. We could again use a separate `JSContext` for this.
 * ZiplineLoader.loadFromManifest() uses `JS_Eval`. We could precompile a function that takes a `mainModuleId` and `mainFunction`.

**Comment by swankjesse** [comment]
One thing that might work – change `QuickJs.evaluate()` to always work in two steps:
1. compile on an eval-carrying `JSContext`
2. execute on an eval-free `JSContext`. 

That way the only eval calls must come from Kotlin, and can’t be internal-to-JS.

**Comment by JakeWharton** [comment]
Sounds like a plan. I was going to suggest something similar but slightly more convoluted.

**Comment by swankjesse** [comment]
Same plan, now in smaller steps:

 * [x] Create two JS_Context instances in QuickJS C++ code.
    `compilerJsContext`: this one only compiles and never executes
    `jsContext`: this one only executes and never evals
 * [x] Change C++ implementation code to avoid calling eval on the regular `jsContext`. Instead it does a two step process of compile and then execute. 
 * [x] Remove `evaluate()` as public API from QuickJS. It needs to remain as an internal API for tests!

**Comment by swankjesse** [comment]
The dangerous bits of `eval()` are gone. These tasks are outstanding for a future release:

 * [ ] Find our (3 or 4) callsites in Zipline that call evaluate and build special-purpose APIs to do what they need exactly. These can initially use evaluate() internally.
 * [ ] Consider using manual QuickJS C calls for evaluate() internal calls in 3.
 * [ ] Don't initialize the compilerJsContext in normal execution paths. This requires something special for finalization-registry such as shipping bytecode!
